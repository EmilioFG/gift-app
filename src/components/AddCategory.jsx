import { useState } from 'react';
import PropTypes from 'prop-types';


export const AddCategory = ({ onNewCategory }) => {

  const [inputValue, setInputValue] = useState('')

  const handleSubmit = (e) => {
    e.preventDefault();
    if (inputValue.trim().length <= 1) return;

    onNewCategory(inputValue.trim());
    setInputValue('');
  }

  const handleChange = ({ target }) => setInputValue(target.value)

  return (
    <>
      <h3>Agregar Categor√≠a</h3>
      <form onSubmit={handleSubmit} aria-label="form">
        <input
          type="text"
          placeholder="Buscar Gif"
          value={inputValue}
          onChange={handleChange}
        />
      </form>
    </>
  )
}

AddCategory.prototypes = {
  onNewCategory: PropTypes.func.isRequired,
}
