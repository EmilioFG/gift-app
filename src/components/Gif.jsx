import PropTypes from 'prop-types';


export const Gif = ({ title, url}) => {
  return (
    <div className="card">
      <img src={url} alt={title} />
      <h6>{title}</h6>
    </div>
  )
}

Gif.prototypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
}
