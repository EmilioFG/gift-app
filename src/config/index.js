export default {
  API_URL: process.env.VITE_APP_API_URL,
  API_KEY: process.env.VITE_APP_API_KEY,
}