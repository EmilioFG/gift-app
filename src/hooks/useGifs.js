import { useState, useEffect } from 'react';

import { getGifs } from '../api/gif';


export const useGifs = (category) => {

  const [gifs, setGifs] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    getApiGif(category);
  }, []);

  const getApiGif = async (category) => {
    const gifs = await getGifs(category);
    setGifs(gifs);
    setIsLoading(false)
  }

  return {
    gifs,
    isLoading,
  }
}