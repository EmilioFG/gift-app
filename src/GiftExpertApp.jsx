import { useState } from 'react'

import { AddCategory, GifGrid } from './components';


export const GiftExpertApp = () => {

  const [categories, setCategories] = useState(['dragon ball z']);

  const handleAddCategory = (newCategory) => {
    const currentCategories = [...categories].map((category) => (category.toLowerCase()));
    if (currentCategories.includes(newCategory.toLowerCase()))
      return alert(`la categor√≠a ${newCategory} ya existe`);

    setCategories([...categories, newCategory]);
  };

  return (
    <>
      <h1>GiftExpertApp</h1>
      <AddCategory onNewCategory={handleAddCategory} />
      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  )
}