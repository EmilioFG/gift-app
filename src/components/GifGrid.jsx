import PropTypes from 'prop-types';

import { useGifs } from '../hooks/useGifs';
import { Gif } from './Gif';


export const GifGrid = ({ category }) => {

  const { gifs, isLoading } = useGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {isLoading && <span>...loading</span>}
      <div className='card-grid'>
          {gifs.map((gif) => (
            <Gif
              key={gif.id}
              {...gif}
            />
          ))}
        </div>
    </>
  )
}

GifGrid.prototypes = {
  category: PropTypes.string,
}
