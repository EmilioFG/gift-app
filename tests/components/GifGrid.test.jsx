import { render, screen } from '@testing-library/react';
import { GifGrid } from '../../src/components';
import { useGifs} from '../../src/hooks/useGifs';

jest.mock('../../src/hooks/useGifs');


describe("Pruebas en <GifGrid />", () => {

  const category = 'zombie';

  test("debe de hacer match con el snapchot", () => {

    useGifs.mockReturnValue({
      gifs: [],
      isLoading: true,
    });

    const { container } = render(<GifGrid category={category} />);
    expect(container).toMatchSnapshot();
  });

  test("debe de mostrar el loading inicial", () => {

    useGifs.mockReturnValue({
      gifs: [],
      isLoading: true,
    });

    render(<GifGrid category={category} />);

    expect(screen.getAllByText('...loading'));
    expect(screen.getByText(category));
  });

  test("debe mostrar gifs cuando se cargan las imÃ¡genes useGifs()", () => {

    const gifs = [
      { id: 'aaa', title: 'zombie', url: 'http://emiliogonzalez.dev'},
      { id: 'bbb', title: 'goku', url: 'http://emiliogonzalez.dev'},
    ];

    useGifs.mockReturnValue({
      gifs,
      isLoading: false,
    });

    render(<GifGrid category={category} />);

    // screen.debug()

  });

});
