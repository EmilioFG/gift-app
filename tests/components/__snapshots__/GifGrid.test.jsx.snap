// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> debe de hacer match con el snapchot 1`] = `
<div>
  <h3>
    zombie
  </h3>
  <span>
    ...loading
  </span>
  <div
    class="card-grid"
  />
</div>
`;
